src/construction/construction.py

```diff
@@ -1,16 +1,243 @@
+                          on=key_cols,
+                          how="outer",
+                          suffixes=("_old", ""),
+                          indicator=True)
+    additions_df = outer_join[(outer_join._merge == "right_only")].drop("_merge", axis=1)
+    additions_df = additions_df[additions_df.columns[~additions_df.columns.str.endswith('_old')]]
+
+    if additions_df.shape[0] > 0:
+        additions_df["is_constructed"] = True
+        additions_df["accept_changes"] = False
+        return additions_df
+    else:
+        construction_logger.info("No additions found.")
+        return None
+
+
+def output_construction_files(amendments_df, additions_df, config, write_csv, run_id):
+    """Save CSVs of amendments and additions for user approval."""
+    # Prepare output paths
+    network_or_hdfs = config["global"]["network_or_hdfs"]
+    paths = config[f"{network_or_hdfs}_paths"]
+    tdate = datetime.now().strftime("%Y-%m-%d")
+    construction_folder = paths["construction_path"]
+    amendments_filename = os.path.join(construction_folder, f"construction_amendments_{tdate}_v{run_id}.csv")
+    additions_filename = os.path.join(construction_folder, f"construction_additions_{tdate}_v{run_id}.csv")
+
+    # Check if the dataframes are empty before writing
+    if amendments_df is not None:
+        write_csv(f"{amendments_filename}", amendments_df)
+    if additions_df is not None:
+        write_csv(f"{additions_filename}", additions_df)
+
+    return True
+
+
+def apply_construction(main_df, config, check_file_exists, read_csv, write_csv, run_id):
+    """Read user-edited construction files and apply them to the main snapshot."""
+    # Prepare filepaths to read from
+    network_or_hdfs = config["global"]["network_or_hdfs"]
+    paths = config[f"{network_or_hdfs}_paths"]
+    amendments_filepath = paths["construction_edit_path"]
+    additions_filepath = paths["construction_add_path"]
+
+    # Check if the construction files exist
+    amendments_exist = check_file_exists(amendments_filepath)
+    additions_exist = check_file_exists(additions_filepath)
+
+    # If each file exists, read it and call the function to apply them
+    if (not amendments_exist) and (not additions_exist):
+        construction_logger.info("No amendments or additions to apply, skipping...")
+        return main_df
+
+    if amendments_exist:
+        try:
+            amendments_df = read_csv(amendments_filepath)
+            constructed_df = apply_amendments(main_df, amendments_df)
+        except pd.errors.EmptyDataError:
+            construction_logger.warning(f"Amendments file {amendments_filepath} is empty, skipping...")
+
+    if additions_exist:
+        try:
+            additions_df = read_csv(additions_filepath)
+            constructed_df = apply_additions(main_df, additions_df)
+        except pd.errors.EmptyDataError:
+            construction_logger.warning(f"Additions file {additions_filepath} is empty, skipping...")
+
```

Comment:
Once the additional records are added, the indexing is causing issues in imputation - this sorted it: constructed_df = constructed_df.sort_values(["reference", "instance"], ascending=[True, True]).reset_index(drop=True)

Reply:
Having relooked at the accepted changes above, this might not be issue

Reply:
Resolving as this seems to be caused by the zero division bug above